pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,356 @@
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go t.handleTaskUpdate(ctx, prop)
+			}
+			time.Sleep(waitForUpdatesRetry)
+		}
+	}
+}
+
+// handleTaskUpdate is processes each task update in a separate goroutine
+func (t *TaskViewManagerImpl) handleTaskUpdate(ctx context.Context, prop types.PropertyChange) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("handleTaskUpdate for prop: %+v", prop)
+	taskInfo, ok := prop.Val.(types.TaskInfo)
+	if !ok {
+		log.Errorf("failed to cast taskInfo for prop: %+v", prop)
+		return
+	}
+	if taskInfo.State == types.TaskInfoStateQueued || taskInfo.State == types.TaskInfoStateRunning {
+		log.Infof("task not completed yet. current state: %v", taskInfo.State)
+		return
+	}
+	result := Result{}
+	t.mu.Lock()
+	requestorChan, err := t.getChannelFromTask(ctx, taskInfo.Task)
+	t.mu.Unlock()
+	if err != nil {
+		result.TaskInfo = nil
+		result.Err = err
+	} else if taskInfo.State == types.TaskInfoStateError {
+		log.Debugf("taskInfo.Error: %+v", taskInfo.Error)
+		result.TaskInfo = nil
+		result.Err = errors.New(taskInfo.Error.LocalizedMessage)
+	} else {
+		result.TaskInfo = &taskInfo
+		result.Err = nil
+	}
+
+	requestorChan <- result
+}
+
+// AddTaskToListView adds task to listView and the internal map
+func (t *TaskViewManagerImpl) AddTaskToListView(ctx context.Context, taskMoRef types.ManagedObjectReference, ch chan Result) error {
+	log := logger.GetLogger(ctx)
+	log.Debugf("AddTaskToListView for %+v", taskMoRef)
+	for t.getListView() == nil {
+		log.Debugf("listview not created. waiting for 0.5 sec")
+		time.Sleep(time.Millisecond * 500)
+	}
+	listView := t.getListView()
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	err := listView.Add(ctx, []types.ManagedObjectReference{taskMoRef})
+	if err != nil {
+		return logger.LogNewErrorf(log, "failed to add task to ListView. error: %+v", err)
+	}
+	log.Debugf("Added task %+v to listView", taskMoRef)
+	t.addTaskToMap(ctx, taskMoRef, ch)
+	return nil
+}
+
+func (t *TaskViewManagerImpl) addTaskToMap(ctx context.Context, taskMoRef types.ManagedObjectReference, ch chan Result) {
+	// t.taskMap[TaskDetails{Reference: taskMoRef}] = ch
+	t.taskMap[taskMoRef] = &TaskDetails{
+		MarkedForRemoval: false,
+		ResultCh:         ch,
+	}
+	log := logger.GetLogger(ctx)
+	log.Infof("Added task %+v and channel to internal map", taskMoRef)
+}
+
+func (t *TaskViewManagerImpl) removeTaskFromMap(ctx context.Context, taskMoRef types.ManagedObjectReference) {
+	// delete(t.taskMap, TaskDetails{Reference: taskMoRef})
+	delete(t.taskMap, taskMoRef)
+	log := logger.GetLogger(ctx)
+	log.Infof("Removed task %+v from internal map", taskMoRef)
+}
+
+func (t *TaskViewManagerImpl) getChannelFromTask(ctx context.Context, taskMoRef types.ManagedObjectReference) (chan Result, error) {
+	log := logger.GetLogger(ctx)
+	// ch, ok := t.taskMap[taskDetails]
+	taskDetails, ok := t.taskMap[taskMoRef]
```

Comment:
we can get the required channel directly using ` t.taskMap[taskMoRef]` in `handleTaskUpdate`.
unless you have the plan to make use of `getChannelFromTask` elsewhere let's not add it as a separate function.